src/JsonReader.cs

```diff
@@ -211,6 +211,108 @@ public static JsonReadResult<T> TryRead<T>(this IJsonReader<T> reader, ReadOnlyS
         }
     }
 
+    public static IAsyncEnumerator<KeyValuePair<string, T>>
+        GetObjectAsyncEnumerator<T>(this IJsonReader<T> reader,
+                                    Stream stream, int initialBufferSize) =>
+        GetObjectAsyncEnumerator(reader, stream, initialBufferSize, CancellationToken.None);
+
+    public static IAsyncEnumerator<KeyValuePair<string, T>>
+        GetObjectAsyncEnumerator<T>(this IJsonReader<T> reader,
+                                    Stream stream, int initialBufferSize,
+                                    CancellationToken cancellationToken)
+    {
+        if (reader is null) throw new ArgumentNullException(nameof(reader));
+        if (stream is null) throw new ArgumentNullException(nameof(stream));
+        if (initialBufferSize < 0) throw new ArgumentOutOfRangeException(nameof(initialBufferSize), initialBufferSize, null);
+
+        return GetObjectAsyncEnumeratorCore(reader, stream, initialBufferSize, cancellationToken);
+    }
+
+    static async IAsyncEnumerator<KeyValuePair<string, T>>
+        GetObjectAsyncEnumeratorCore<T>(IJsonReader<T> reader,
+                                        Stream stream, int initialBufferSize,
+                                        CancellationToken cancellationToken)
+    {
+        using var scr = new StreamChunkReader(stream, initialBufferSize);
+
+        var state = new JsonReaderState();
+        var sm = new ObjectReadStateMachine();
+        ObjectReadStateMachine.ReadResult readResult;
+        string? name = null;
+
+        do
+        {
+            var readTask = scr.ReadAsync(cancellationToken);
+            if (!readTask.IsCompleted)
+                await readTask.AsTask().ConfigureAwait(false);
+
+            while (true)
+            {
+                var read = TryReadItem(scr.RemainingChunkSpan, out var bytesConsumed, out readResult, out var item);
+                scr.ConsumeChunkBy(bytesConsumed);
+                if (!read)
+                    break;
+                cancellationToken.ThrowIfCancellationRequested();
+                Debug.Assert(name is not null);
+                yield return KeyValuePair.Create(name, item!);
```

Comment:
[nit] Consider using `Debug.Assert` instead of `item!` as well.

Reply:
Good catch. Done with 7be56c0690c9fd81634c23a8d6d74da38c76c0c1.